similarity_api/src/sim/compute.py

```diff
@@ -108,39 +108,60 @@ def calculate_batch_similarity(batch, nlp):
     return sim_scores
 
 
-def calculate_goal_similarity(goals_list: List[str], goal_ids_list: List[int], grants_list: List[int], alpha: float, batch_size: int = 500) -> List[Dict[str, str]]:
+def calculate_goal_similarity(goals_list: List[str], goal_ids_list: List[int], grants_list: List[int], alpha: float, cluster: bool = False, batch_size: int = 500) -> List[Dict[str, str]]:
     num_goals = len(goals_list)
     matched_goals = []
+    matched_ids = []
 
     for i in range(0, num_goals, batch_size):
         end = min(i + batch_size, num_goals)
         current_batch = goals_list[i:end]
         current_ids = goal_ids_list[i:end]
-        current_recipients = grants_list[i:end]
+        current_grants = grants_list[i:end]
 
         batch_sim_scores = calculate_batch_similarity(current_batch, nlp)
 
-        for j, (goal, goal_id, recipient_id) in enumerate(zip(current_batch, current_ids, current_recipients)):
+        for j, (goal, goal_id, grant_id) in enumerate(zip(current_batch, current_ids, current_grants)):
             cur_sim_scores = batch_sim_scores[j]
             cur_sim_scores[j] = 0
             cur_potential_idx = [k for k, v in enumerate(cur_sim_scores) if v > alpha]
 
-            for k in cur_potential_idx:
-                if i + j > i + k:
-                    continue
-                matched_goals.append({
-                    "goal1": {
-                        "id": goal_id,
-                        "name": goal,
-                        "grantId": recipient_id,
-                    },
-                    "goal2": {
+            if cluster:
+                matches = []
+                for k in cur_potential_idx:
+                    if i + j > i + k or current_ids[k] in matched_ids:
+                        continue
+                    matches.append({
```

Comment:
Can we also include the "isCurated" column in the initial query to the Goals table and only allow one "isCurated" goal in this?

Reply:
yeah, I can add an `isCurated` property based on `GoalTemplate.creationMethod === 'Curated'`. can you explain the second part of your question a bit more? I don't think I understand it completely

Reply:
show all the matches, but if the matches include more than one curated goal, only show one curated goal? is that right?

Reply:
A group of matched goals can only include one "curated" goal, regardless of text match

Reply:
yes, exactly

Reply:
@nvms you can disregard this, unless you've already gone far down the rabbit hole. I had a brainstorm last night about a different way to solve this problem.